### PR TITLE
Tm 4584 UI projected vacancies page and search

### PR DESCRIPTION
**Pull request by cagbakwu**
[Ticket] https://metaphase.atlassian.net/browse/TM-4584
STILL WORKING ON, this is the first half. I have to update the results card info.

-Created projected vacancy search UI page
-Updated drop down options & content
-Added redux files for  search section
-Removed some of my previous lodash get calls.

Screenshot 2023-06-30 at 3 47 51 PM

### CODE REVIEW

**Comment by pveland621** [comment]
Menu styling is a little off on the dashboard landing page:
image

This is definitely the longest (by character) menu option we have, so maybe just have a hanging indent:
image

It does fix itself when you go to the PV page though - but even then we might need a tiny bit of padding on the right:
image

**Comment by pveland621** [comment]
Weirdly, at different resolutions, the label and dropdown is spaced wrong by just a bit:
image

And again, super tiny nitpick but on 1100px resolution, spacing is a little off:
image

Spacing @ 1600px:
image

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+
+  const userSelections = useSelector(state => state.editProjectedVacancy);
+  const dummyPositionDetails = useSelector(state => state.editProjectedVacancy);
+  const [limit, setLimit] = useState(get(userSelections, 'limit') || EDIT_POSITION_DETAILS_PAGE_SIZES.defaultSize);
+  const [ordering, setOrdering] = useState(get(userSelections, 'ordering') || EDIT_POSITION_DETAILS_SORT.defaultSort);
+
+  const genericFiltersIsLoading = useSelector(state => state.filtersIsLoading);
+  const genericFilters = useSelector(state => state.filters);
+
+  const [selectedBureaus, setSelectedBureaus] = useState(get(userSelections, 'selectedBureaus') || []);
```
Comment (pveland621) [inline src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:33]: There's no way you could've known, but we are trying to move away from using lodash, specifically lodash `get`. Instead, we are trying to move the project towards using [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining). It'd be a ton of refactoring for this component so I'll defer to other devs on whether or not you should spend the time just changing the `get`s to optional chaining.

File: src/actions/projectedVacancy.js [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+export function editProjectedVacancyFetchData() {
+  return (dispatch) => {
+    batch(() => {
+      dispatch(editProjectedVacancyFetchDataSuccess(dummyPositionDetails));
+      dispatch(editProjectedVacancyFetchDataLoading(false));
+      dispatch(editProjectedVacancyFetchDataErrored(false));
+    });
+  };
+}
```
Comment (pveland621) [inline src/actions/projectedVacancy.js:209]: awesome job on this but one tiny thing is we do want to send off the `editProjectedVacancyFetchDataLoading` as the last dispatch in the batch, as we use the loading state to determine when to render (like you did in this new component: 
image
we always want the loading dispatch to be the first (when setting loading to true) and last (when setting loading to false) dispatches in the action, or as elizabeth likes to call it, a `loading sandwich`:
image

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+  }
+
+  const throttledTextInput = () =>
+    throttle(q => setTextInput(q), 300, { leading: false, trailing: true });
+
+  const setTextInputThrottled = (q) => {
+    throttledTextInput(q);
+  };
+
+  function renderSelectionList({ items, selected, ...rest }) {
```
Comment (pveland621) [inline src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:153]: last thing, I noticed this is based on the `PanelMeetingAgendas` filter setup - i saw in the PR description that you still need to work on the dropdown options and content, just making sure that this is what you meant you needed to fix

File: src/actions/projectedVacancy.js (hunk 2 above)
Comment (cagbakwu) [reply src/actions/projectedVacancy.js:209]: ^^^ these visuals are great! @pveland621

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx (hunk 1 above)
Comment (cagbakwu) [reply src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:33]: @pveland621 @M00staff  ok that sounds good, possibly could wait if no one has any objections? If everything is complete before end of sprint, I could go back through and refactor?

Comment (M00staff) [reply src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:33]: its actually not super difficult to swap - this line specifically would just look something like this:

`const [selectedBureaus, setSelectedBureaus] = useState(userSelections?.selectedBureaus || []);`

its one of the reasons I'm so vehemently anti-lodash lol - everything except maybe checking for deeply nested equality has an easy vanilla solution, at a fraction of the load time.

.. .all that being said - in the interest of time - and seeing as how we don't have it as a luxury - am 100% cool with keeping all the lodash here, as is

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+                  </div>
+                  <div className="filter-div">
+                    <div className="label">Bid Cycle:</div>
+                    <Picky
+                      {...pickyProps}
+                      placeholder="Select a Bid Cycle"
+                      value={selectedBidCycle}
+                      options={cycleOptions}
+                      onChange={setSelectedBidCycle}
+                      valueKey="code"
```
Comment (M00staff) [inline src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:316]: Bid Cycles is being wonky - when you click one, it auto clicks all of them ... i think this should be `id` but there also might be something else amiss 🤔

image

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+    // Free Text
+    q: textInput || textSearch,
+  });
+
+  const resetFilters = () => {
+    setSelectedBureaus([]);
+    setSelectedLocations([]);
+    childRef.current.clearText();
+    setClearFilters(false);
+  };
```
Comment (M00staff) [inline src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:91]: you're gonna wanna clear our the rest of the filters here - i also had mine clear the search box ... not sure if that's right - but it made sense to me

image

File: src/Components/ProfileMenu/NavLink/NavLink.jsx [hunk 6]
```diff
@@ -96,7 +96,8 @@ class NavLink extends Component {
             <div className="list-item-wrapper">
               {
                 this.wrapInLink( // wrap our element
-                  <span>
+                /* this flex keeps our icon and title on the same line */
+                  <span style={{ display: 'flex' }}>
```
Comment (M00staff) [inline src/Components/ProfileMenu/NavLink/NavLink.jsx:100]: HAH! 

in-line styles - i miss those!

so handy

Comment (M00staff) [reply src/Components/ProfileMenu/NavLink/NavLink.jsx:100]: though - it looks like this breaks a little when the card gets expanded ... it might be that the new card we end up using doesn't have this issue 🤔
image

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+    selectedBureaus,
+    selectedOrgs,
+    selectedGrade,
+    selectedLanguage,
+    selectedSkills,
+    textSearch,
+  });
+
+  useEffect(() => {
+    dispatch(editProjectedVacancyFiltersFetchData(getQuery()));
```
Comment (M00staff) [inline src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:103]: so this will run the Query on page load when there are likely no filters ... i dont *THINK* that is expected behavior

i could be wrong on that though

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+      selectedLanguage,
+      selectedSkills,
+      textSearch,
+    ];
+    if (isEmpty(filter(flatten(filters))) && isEmpty(textSearch)) {
+      setClearFilters(false);
+    } else {
+      setClearFilters(true);
+    }
+    dispatch(editProjectedVacancyFetchData(getQuery()));
```
Comment (M00staff) [inline src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:123]: and this dispatch will run the fetch query on the data every time you change a filter ... the way i had mine setup was to only run the `fetchDataQuery` when the submit button was pressed. 

i know `PanelMeetingAgendas` and `EmployeeAgendaSearch` and `Position Manager` all work a little differently from that (and each other)

.. might need more clarification on how this is supposed to function 🤔

File: src/Components/ProfileMenu/NavLink/NavLink.jsx (hunk 6 above)
Comment (cagbakwu) [reply src/Components/ProfileMenu/NavLink/NavLink.jsx:100]: @M00staff I added some CSS "Magic" to resolve this.

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx (hunk 5 above)
Comment (cagbakwu) [reply src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:91]: @M00staff cleared!

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx (hunk 4 above)
Comment (cagbakwu) [reply src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:316]: @M00staff this was weird but its resolved now

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx (hunk 8 above)
Comment (cagbakwu) [reply src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:129]: @M00staff if this changes, I can revise it. For now I can leave it as is.

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+  const post = genericFilters$.find(f => get(f, 'item.description') === 'post');
+  const locationOptions = uniqBy(sortBy(get(post, 'data'), [(p) => p.city]), 'code');
+  const grades = genericFilters$.find(f => get(f, 'item.description') === 'grade');
+  const gradesOptions = uniqBy(get(grades, 'data'), 'code');
+  const skills = genericFilters$.find(f => get(f, 'item.description') === 'skill');
+  const skillsOptions = uniqBy(sortBy(get(skills, 'data'), [(s) => s.description]), 'code');
+  const languages = genericFilters$.find(f => get(f, 'item.description') === 'language');
+  const languagesOptions = uniqBy(sortBy(get(languages, 'data'), [(c) => c.custom_description]), 'custom_description');
+  const cycles = genericFilters$.find(f => get(f, 'item.description') === 'bidCycle');
+  const cycleOptions = uniqBy(sortBy(get(cycles, 'data'), [(c) => c.custom_description]), 'custom_description');
```
Comment (jlam-mpc) [inline src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:54]: Some of these are written plural and some are singular (same goes for the mockups) while the functionality of the filters is multi-select for all of these. Is it confirmed that we can select multiple for all of these fields or do we need to convert some of them to radio buttons and non-array types?

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+    dropdownHeight: 255,
+    renderList: renderSelectionList,
+    includeSelectAll: true,
+  };
+
+  const dummyid = get(dummyPositionDetails, 'id', '');
+
+  return (
+    isLoading ?
+      <Spinner type="projected-vacancy-filters" size="small" /> :
```
Comment (jlam-mpc) [inline src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:207]: While the page is close to perfectly styled, I have some suggestions on the organization of it to reduce css bloat and reuse some common style blocks throughout the app. Position Manager Page, Publishable Position Page, and Projected Vacancy Page (potentially more) all use very similar styles. In my Publishable Positions Page, I'm reusing most of the styles in _bureau (following Position Manager Page) and added an umbrella className to narrow down to specifics. I think it'd be good to eventually refactor all 3 components to have generic classes with as much of the same style declarations as possible.

TLDR;
There is some mild CSS redundancy, you should leave the styling you have for now, we should refactor the styles/classes for this component after it's been merged to align with Position Manager and Publishable Positions

Comment (cagbakwu) [reply src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:207]: @jlam-mpc yea I noticed the same stylings as well. I think refactoring the css afterwards is a great idea.

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx (hunk 9 above)
Comment (cagbakwu) [reply src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:54]: @jlam-mpc from the other search pages, they all have multi selects and from the screenshots provided. I could verify if they want to change or not.

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+
+  const fetchAndSet = () => {
+    const filters = [
+      selectedBureaus,
+      selectedPost,
+      selectedOrgs,
+      selectedGrade,
+      selectedLanguage,
+      selectedSkills,
+      selectedBidCycle,
```
Comment (M00staff) [inline src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:122]: just an FYI - this guy was missing in a few places so it wouldn't trigger the 'clear filters' button, or be sent to the action ... i just added it, so you should pull

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx (hunk 8 above)
Comment (M00staff) [reply src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:130]: yea - based on the slack convo, i think we're good here

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx [hunk 12]
```diff
@@ -0,0 +1,379 @@
...
+    setSelectedBidCycle([]);
+    setTextSearch('');
+    setTextInput('');
+    childRef.current.clearText();
+    setClearFilters(false);
+  };
+
+  const getCurrentInputs = () => ({
+    selectedBureaus,
+    selectedPost,
```
Comment (M00staff) [inline src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:101]: this was also missing - just added it

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx (hunk 3 above)
Comment (M00staff) [reply src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:162]: actually - dont worry about trimming this down, ill abstract it all in another PR

File: src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx (hunk 12 above)
Comment (cagbakwu) [reply src/Components/BureauPage/ProjectedVacancySearch/EditProjectedVacancyDetails.jsx:101]: @M00staff ah thanks, I thought I added them but it probably got lost amongst the other calls haha.

**Review by M00staff — APPROVED** [review]
This should be OK to merge in - no need to worry about updating the card, as that will be in another ticket.

A few things though:

I did a bunch of renaming - there will be a separate ticket that handles editing the vacancies that this search page produces ... the naming might have gotten a little complicated at that point (edit was used all over the place, even though we aren't editing anything, yet)... make sure to pull and click through and verify I didn't bonk anything

Also,
image

I don't **think** this ^^ set of actions & reducers are doing anything, or used anywhere... you get generic filter options to use from `filtersFetchData`, and when they are selected and stored in Redux - they are stored in the `projectedVacancySelections` reducer.

Unless you were planning on using them for something (or if I'm off) get rid of them.

**Review by M00staff — CHANGES_REQUESTED** [review]
actually scratch that - there is something wonky with your test suite
image
it should actually be failing at the `server` step ... there's probably a typo somewhere

**Review by M00staff — APPROVED** [review]
